python_modules/dagster/dagster/_core/definitions/data_version.py

```diff
@@ -450,24 +454,41 @@ def asset_graph(self) -> "AssetGraph":
             self._asset_graph = check.not_none(self._asset_graph_load_fn)()
         return self._asset_graph
 
+    # This is lazily constructed because it depends on the asset graph, which needs to be lazily
+    # constructed for GQL performance reasons.
+    @property
+    def instance_queryer(self) -> "CachingInstanceQueryer":
+        from dagster._utils.caching_instance_queryer import CachingInstanceQueryer
+
+        if self._instance_queryer is None:
+            self._instance_queryer = CachingInstanceQueryer(self._instance, self.asset_graph)
+        return self._instance_queryer
+
     @cached_method
     def _get_current_data_version(self, *, key: AssetKey) -> DataVersion:
-        is_source = self.asset_graph.is_source(key)
-        event = self._instance.get_latest_data_version_record(
-            key,
-            is_source,
-        )
-        if event is None and is_source:
-            return DEFAULT_DATA_VERSION
-        elif event is None:
-            return NULL_DATA_VERSION
+        # Currently we can only use asset records, which are fetched in one shot, for non-source
+        # assets. This is because the most recent AssetObservation is not stored on the AssetRecord.
+        if self.asset_graph.is_source(key):
+            event = self._instance.get_latest_data_version_record(
```

Comment:
this can be replaced with a `self.instance_queryer.get_latest_materialization_or_observation_record()`, which handles reading from the asset record cache for you. The benefit here is that if we plan on storing observations in the record cache, this code can immediately take advantage of that performance boost